Allow use of DiscreteUniform distribution for CategoricalGibbsMetropolis step method. #3419

helmutsimon opened the pull request:
The CategoricalGibbsMetropolis step method is the default for the Categorical distribution. The Metropolis step method may not work for Categorical in some circumstances, e,.g. as part of compound distribution. However, CategoricalGibbsMetropolis currently can only be used with Categorical or Bernoulli distributions.

For some models it is necessary to use a DiscreteUniform distribution rather than Categorical, as the k parameter can be very large. This change allows the CategoricalGibbsMetropolis step method to be used for DiscreteUniform distributions, provided that the parameter `lower' is zero.

Commit:
Update metropolis.py

Allow CategoricalGibbsMetropolis method for DiscreteUniform distributions (when parameter `lower' ==0).

Commit:
Merge pull request #1 from helmutsimon/helmutsimon-categoricalgibbsmetropolis-1

Allow use of DiscreteUniform distribution for CategoricalGibbsMetropolis step method. #3419

lucianopaz commented on pymc3/step_methods/metropolis.py:
```diff
@@ -400,9 +400,14 @@ def __init__(self, vars, proposal='uniform', order='random', model=None):
                 k = 2
+            elif isinstance(distr, pm.DiscreteUniform):
+                k = draw_values([distr.upper])[0]
+                if draw_values([distr.lower])[0] != 0:
```
Draw values of `lower` could return an array, and not a scalar. You should take this into account and wrap the equality comparison in an `np.all`

helmutsimon replied:
@lucianopaz I have amended the change and also added PRs for examples and Release Notes. This is my first PR, so I appreciate your assistance with this.

helmutsimon replied:
@lucianopaz I have amended the test code to improve coverage. The travis tests seem to have been affected by matplotlib issue #3443.

lucianopaz reviewed (changes requested):
Could you add a test to the test suite that verifies that the step method works well with the DiscreteUniform distribution of lower is 0?

Also, you should add a line about this fix in the release notes for under maintenance.

Commit:
Update metropolis.py/CategoricalGibbsMetropolis(ArrayStep)

Add change to PR #3435 as requested by reviewer. In testing that parameter 'lower' is zero, allow for possibility that it may be an array.

helmutsimon commented:
To be replaced with new pull request.
